cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,114 +285,675 @@ func copyFromGlobHelper(src, dst, globEx string) error {
 	return nil
 }
 
-func generateMountsModeRecover(mst *initramfsMountsState) error {
-	// steps 1 and 2 are shared with install mode
-	if err := generateMountsCommonInstallRecover(mst); err != nil {
-		return err
+// recoverDegradedState is a map which usually contains the following keys:
+// - "ubuntu-data" - a map of states for the partition including the keys below
+// - "ubuntu-boot" - same as ubuntu-data
+// - "ubuntu-save" - same as ubuntu-data
+// - "error-log" - a list of errors encountered while setting up recover mode
+//
+// partitions can have the following keys in their maps:
+// - "mount-state" - whether it was mounted successfully or not
+// - "mount-location" - where this partition was mounted
+// - "device" - what device the partition corresponds to
+// - "locate-state" - whether the partition on the disk was located or not
+// - "unlock-state" - whether it was unlocked successfully or not
+// - "unlock-key" - what key it was unlocked with (TODO: should this be rolled into "unlock-state" ?)
+type recoverDegradedState map[string]interface{}
```

Comment:
there should be a way along the lines of some of the helpers in asserts/ifacedecls.go  to make this again a struct of structs but keep some level of generic setting support

```
type recoverDegradedState struct {
   UbuntuData  partitionState ...
   ErrorLog []string  ...
}

func (r *recoverDegradedState) partition(part string) *partitionState {
   switch part {
   case "ubuntu-data":
        return &r.UbuntuData
        ...
   }
}
```

etc

Reply:
ok, understood